src/dcpu16n/dcpu16n.cpp

```diff
@@ -0,0 +1,1047 @@
+            if( (opcl & 0x001f) != 0 ) {
+                skip = DCPU16N_skipstateN[ opcl & 0x001f];
+                pc  += DCPU16N_skipadd[ (opcl >> 10) & 0x3f ];
+                pc  += DCPU16N_skipadd[ (opcl >>  5) & 0x1f ];
+            }
+            else if((opcl & 0x03e0) != 0) {
+                skip = false;
+                pc  += DCPU16N_skipadd[ (opcl >> 10) & 0x3f ];
+            }
+            else {
+                skip = DCPU16N_skipstateI[ (opcl >> 10) & 0x001f ];
+            }
+            phase = DCPU16N_PHASE_OPFETCH;
+            break;
+
+        case DCPU16N_PHASE_EXECJMP:
+            sp -= 2;
+            bca = emu[(sp >> 12) & 0xf] | (sp & 0x0fff);
+            vcomp->WriteB(bca, (Byte)(bcu & 0x0ff));
+            vcomp->WriteB(bca + 1, (Byte)(bcu >> 8));
+            pc = acu;
+            phase = DCPU16N_PHASE_OPFETCH;
+            break;
+
+        case DCPU16N_PHASE_EXECRFI:
+            phase = DCPU16N_PHASE_OPFETCH;
+            qint  = false;
+
+            // pop A
+            bca   = emu[(sp >> 12) & 0xf] | (sp & 0x0fff);
+            sp   += 2;
+            r[0]  = ( ( (Word)vcomp->ReadB(bca + 1) ) << 8 )
+                      | (Word)vcomp->ReadB(bca);
+
+            // pop PC
+            bca   = emu[(sp >> 12) & 0xf] | (sp & 0x0fff);
+            sp   += 2;
+            pc    = ( ( (Word)vcomp->ReadB(bca + 1) ) << 8 )
+                      | (Word)vcomp->ReadB(bca);
+            break;
+
+        case DCPU16N_PHASE_MARKSKIP:
+            skip = true;
+            phase = DCPU16N_PHASE_OPFETCH;
+            break;
+
+        case DCPU16N_PHASE_INTERRUPT:
+            phase = DCPU16N_PHASE_OPFETCH;
+            qint  = true; // Queue interrupts
+
+            // push PC
+            sp   -= 2;
+            bca   = emu[(sp >> 12) & 0xf] | (sp & 0x0fff);
+            vcomp->WriteB(bca, (Byte)(pc & 0x0ff));
+            vcomp->WriteB(bca + 1, (Byte)(pc >> 8));
+
+            // push A
+            sp -= 2;
+            bca = emu[(sp >> 12) & 0xf] | (sp & 0x0fff);
+            vcomp->WriteB(bca, (Byte)(r[0] & 0x0ff));
+            vcomp->WriteB(bca + 1, (Byte)(r[0] >> 8));
+
+            // set interrupt address and message
+            pc = ia;
+            r[0] = intq[iqe]; // remove one from queue
+            iqe++;
+            iqc--;
+            if(iqe > 255)
+                iqe = 0;
+            break;
+
+        default:
+            phase = 0;
+            break;
+        }
+        pwrdraw += 5;
+    }
+}
+
+bool DCPU16N::SendInterrupt(Word msg)
+{
+    if(ia == 0)
+        return true;
+
+    if(iqc > 256) {
+        fire = true;
+        return false;
+    }
+    iqc++;
+    intq[iqp] = msg;
+    iqp++;
+    if(iqp > 255)
+        iqp = 0;
+    return true;
+}
+
+Word DCPU16N::IORead(Word addr)
+{
+    return vcomp->ReadW(0x00110000 | addr);
+}
+
+void DCPU16N::IOWrite(Word addr, Word v)
+{
+    vcomp->WriteW(0x00110000 | addr, v);
+}
+
+bool DCPU16N::DoesTrap(Word& msg) {
+	return false;
```

Comment:
I wasn't certain the expected behavior of DoesTrap, @Meisaka can you confirm what it should be?

Reply:
from what I can tell, if this returns true, it will cause an interrupt to get sent back to the CPU towards the end of a "tick".
but since the 16N has a built-in queue, and my implementation used a crazy phase state machine, it's not really needed.
**tl;dr: just return false;**